README.md
# House of Isometry
House of Isometry is a jQuery plugin that help to create a building in a 3d isometric projection from basic DOM element, all positioned in absolute.
You can navigate through an exploded view of the rooms which are all interactive. They will give you some more information about them & their content.

![House of Isometry](documentation/houseiso.png)

This is a project made with ♥️  by Antistatique a Swiss web agency based in Lausanne for one of our client. The amazing illustration are from Philippe Mignotte

## Use the library
- Clone the project

- Embed a version of jQuery
```
<script src="https://code.jquery.com/jquery-3.2.0.min.js" integrity="sha256-JAW99MJVpJBGcbzEuXk4Az05s/XyDdBomFqNlM3ic+I=" crossorigin="anonymous"></script>
```
- Include the following assets in your project :
```
<link rel="stylesheet" type="text/css" href="house-of-isometry.css">
<script src="house-of-isometry.bundle.js"></script>
```
- You can then initalize the plugin with:
```
var isometricLayout = new IsometricGrid($('.isometric-layout'), 200, 82, 30);
```


## Methods
### showExpandedView()
Show the expanded view of the grid
```
  isometricLayout.showExpandedView();
```
### showHomeView()
Show the home view of the grid
```
  isometricLayout.showHomeView();
```

### reCenter()
Relayout the grid. Usefull when you resize the viewport.
$(window).resize(function() {
  isometricLayout.reCenter();
});
## Build the library
### Javascript

To build your own distribution need the following NPM

Then you need to install all the npm dependencies
```
$ npm install
```
or (when using yarn)
```
$ yarn
```
You can now use the following command
```
$ npm run build
$ npm run watch
```
or more specifically
```
$ npm run build-js
$ npm run build-css
$ npm run watch-js
$ npm run watch-css
```

The sources are situated in `/src`.

changelog.md
## 1.0.0 - 2017-06-27
* Initial release.

gulpfile.babel.js
import gulp from 'gulp';
import autoprefixer from 'autoprefixer';
import reporter from 'postcss-reporter';
import loadPlugins from 'gulp-load-plugins';
import webpack from 'webpack-stream';
import sourcemaps from 'gulp-sourcemaps';
import rename from 'gulp-rename';
import uglify from 'gulp-uglify';
import util from 'gulp-util';


const $ = loadPlugins();

function errorAlert(error) {
  $.notify.onError({ title: 'SCSS Error', message: 'Check your terminal', sound: 'Sosumi', })(error);
  $.util.log(error.messageFormatted ? error.messageFormatted : error.message);
  this.emit('end');
}

/**
 * Build scripts with WebPack
 * With error reporting on compiling (so that there's no crash)
 */
gulp.task( "scripts", function() {
   return gulp.src( "src/house-of-isometric.js" )
      .pipe( gulp.dest( "./dist" ) )
      .pipe( sourcemaps.init( { loadMaps: true } ) )
      .pipe(uglify().on('error', function(e){
        console.log(e);
      }))
      .pipe( rename( "house-of-isometric.min.js" ) )
      .pipe( sourcemaps.write( "./" ) )
      .pipe( gulp.dest( "dist/" ) );
} );

/**
 * Build styles from SCSS files
 * With error reporting on compiling (so that there's no crash)
 */
gulp.task('stylesBuild', () => {
  return gulp.src([`src/house-of-isometric.scss`])
    .pipe($.plumber({ errorHandler: errorAlert }))
    .pipe($.util.noop())
    .pipe($.sass({
      outputStyle: 'compressed',
      precision: 5,
      includePaths: ['.'],
    }))
    .pipe($.postcss([
      autoprefixer({
        browsers: ["last 2 versions", "safari 5", "ie 8", "ie 9", "ff 27", "opera 12.1"],
        options: {
          map: true,
        }
      })
    ]))
    .pipe($.util.noop())
    .pipe($.cleanCss())
    .pipe($.concat('main.css'))
    .pipe($.size({ title: 'STYLES', showFiles: true }))
    .pipe(gulp.dest(`dist`));
});

gulp.task('stylesLint', () => {
  return gulp.src([`src/**/*.s+(a|c)ss`])
      .pipe($.plumber({ errorHandler: errorAlert }))
      .pipe($.postcss(
        [
          require('stylelint')(),
          reporter({
            clearMessages: true,
            throwError: true
          })
        ],
        {
          syntax: require('postcss-scss')
        }));
});

const styles = gulp.task('styles', ['stylesLint', 'stylesBuild']);


webpack.config.js
/* globals require, module, __dirname */

const webpack = require('webpack');
const path = require('path');

module.exports = {
  devtool: 'hidden-source-map',
  context: path.join(__dirname, './src'),
  entry: {
    app: './index.js'
  },
  output: {
    path: path.join(__dirname, 'dist/'),
    filename: '[name].bundle.js',
  },
  externals: {
    'jquery': 'jQuery',
  },
  module: {
    loaders: [
      {
        test: /\.(js|jsx)$/,
        exclude: /(node_modules|bower_components)/,
        loaders: ['babel?presets[]=es2015,plugins[]=transform-es2015-spread,plugins[]=transform-object-rest-spread']
      },
      {
        test: /\.json$/,
        loader: 'json',
      },
      {
        test: /\.html$/,
        loader: 'file',
      },
    ],
  },
  resolve: {
    extensions: ['', '.js', '.jsx'],
    modules: [
      path.resolve('./app'),
      'node_modules',
    ],
    alias: {
      L: 'leaflet',
      ClipperLib: 'clipper-lib',
      R: 'ramda',
    },
  },
  plugins: [

  ],
};

src/house-of-isometry.js
/* global $ */

class Block {
  /**
    * @desc Constructor of the block class
    * @param {JQuery} grid - Grid object containing the block
    * @param {JQuery} element - Element object of the DOM block
  */
  constructor(grid, element) {
    this.grid = grid;
    this.element = element;
    this.currentcoordinates = JSON.parse(element.attr('data-coordinates'));
    this.element
      .on('mousedown touchstart', (e) => {
        e.preventDefault();
        this.grid.clickedElement = $(e.currentTarget);
      })
      .on('click touch', () => {
        if (!grid.expanded) {
          this.grid.showExpandedView();
        }
      });
  }

  /**
    * @desc Return the position of the element from his current coordinates
    * @return {integer[]} - An array with the X/Y current coordinates.
  */
  getPosition() {
    const x = ((this.currentcoordinates[0] - this.currentcoordinates[1]) *
      (this.grid.tileWidth / 2)) + (this.grid.center[0] - (this.grid.tileWidth / 2));
    const y = ((this.currentcoordinates[0] + this.currentcoordinates[1]) *
      (this.grid.tile_height / 2)) + (this.grid.center[1] - (this.grid.tile_height / 2));
    return [x, y];
  }
}

/* eslint-disable no-unused-vars*/
/*
 * Disable the linter for the issue when a class is defined but never used
 * because IsometricGrid in used externally.
*/
class IsometricGrid {
/* eslint-enable */
  /**
    * @desc Constructor of the Isometric Grid object
    * @param {JQuery} container - The container containing the grid
    * @param {integer} tileWidth - The width of a isometric tile width (or horizontal diagonale)
    * @param {integer} tileWidth - The width of a isometric tile heigt (or vertical diagonale)
  */
  constructor(container, tileWidth, floorHeight, gridAngle = 30) {
    const angleRadiant = (gridAngle * Math.PI) / 180;

    this.blocks = [];
    this.container = container;
    this.element = this.container.find('.isometric-grid');
    this.center = [parseInt(this.element.width() / 2, 10), 290];
    this.expanded = false;
    this.tileWidth = tileWidth;
    this.tile_height = 2 * (Math.tan(angleRadiant) * (this.tileWidth / 2));
    this.floorHeight = floorHeight;
    this.dragging = false;
    this.disableClick = false;
    this.mouse_start_x = 0;
    this.mouse_start_y = 0;
    this.mouse_last_x = 0;
    this.mouse_last_y = 0;
    this.clickedElement = [];

    // Keep a reference of this object to be accessed inside jQuery call
    const self = this;

    // Find all the block children in the grid and create a Block object with it
    container.find('.isometric-block').each((index, element) => {
      self.addBlock(new Block(self, $(element)));
    });

    // If the grid container already has the expanded class. Set the variable so.
    if (self.container.hasClass('expanded-view')) {
      self.expanded = true;
    }

    // Update the layout
    self.update();

    // As the init position of the block is ready,
    // set the ready class so the CSS can display the grid
    self.container.addClass('ready');

    // Eventlistener for all touch event related to the drag/drop.
    container.on('mousedown touchstart', (e) => {
      self.startDrag(e);
    });
    $(window).on('mousemove touchmove', (e) => {
      self.drag(e);
    });
    $(window).on('mouseup touchend', () => {
      self.endDrag();
    });

    // On detail view: show the expanded view when clicked everywhere on page...
    container
      .find('.information-container-detail, .focus-background')
      .on('click touch', () => {
        self.showExpandedView();
      });

    // .. expect in the card container.
    container.find('.content-detail-container').click((event) => {
      event.stopPropagation();
    });
  }

  /**
    * @desc Convert a pointerEvent event to his X/Y coordinates
    * @param {event} e - The event containing the click & all his attributes
    * @return {array} out - An aray containing the X/Y cliick coord in pixel
  */
  pointerEventToXY(e) {
    this.out = { x: 0, y: 0 };
    if (
      e.type === 'touchstart' ||
      e.type === 'touchmove' ||
      e.type === 'touchend' ||
      e.type === 'touchcancel'
    ) {
      const touch =
        e.originalEvent.touches[0] || e.originalEvent.changedTouches[0];
      this.out.x = touch.pageX;
      this.out.y = touch.pageY;
    } else if (
      e.type === 'mousedown' ||
      e.type === 'mouseup' ||
      e.type === 'mousemove' ||
      e.type === 'mouseover' ||
      e.type === 'mouseout' ||
      e.type === 'mouseenter' ||
      e.type === 'mouseleave'
    ) {
      this.out.x = e.pageX;
      this.out.y = e.pageY;
    }
    return this.out;
  }

  /**
    * @desc Add a block to the grid blocks array.
    * @param {Block} block - The block object to be passed as children
  */
  addBlock(block) {
    this.blocks.push(block);
  }

  /**
    * @desc Update the whole grid layout. Useful after a change of state / view
  */
  update() {
    for (const block of this.blocks) {
      let floorOffset = 0;
      // Adapt layout depending of the state expanded or not.
      if (this.expanded) {
        // Check if the expanded attribute is correct
        const attributedExpanded = block.element.attr('data-coordinates-expanded');
        if ((typeof attributedExpanded === typeof undefined
          || attributedExpanded === false)
          || attributedExpanded === '[]') {
          // If it hasn't any attribute, hide the block
          block.element.addClass('hidden');
        } else {
          // Else, convert the attributes to an object and pass them as current.
          block.currentcoordinates = JSON.parse(attributedExpanded);
        }
      } else {
        // This will show hidden block & reset the original layout.
        block.element.removeClass('hidden');
        block.currentcoordinates = JSON.parse(
          block.element.attr('data-coordinates')
        );
        // Adds the offset according to the floor.
        floorOffset = parseInt(block.element.attr('data-floor'), 10) * this.floorHeight;
      }

      // Get the current position and update the css position with it.
      const currentPosition = block.getPosition();

      block.element.css({
        left: currentPosition[0],
        top: currentPosition[1] - floorOffset,
      });
    }
  }

  /**
    * @desc When the user starts to drag the container
    * @param {event} e - Click event
  */
  startDrag(e) {
    if (
      this.expanded &&
      this.dragging === false &&
      this.container.hasClass('expanded-view')
    ) {
      e.preventDefault();
      this.mouse_start_x = this.mouse_last_x = this.pointerEventToXY(e).x;
      this.mouse_start_y = this.mouse_last_y = this.pointerEventToXY(e).y;
      this.dragging = true;
    }
  }

  /**
    * @desc When the user is dragging the container
    * @param {event} e - Click event
  */
  drag(e) {
    if (this.expanded && this.dragging === true) {
      const newX =
        (this.pointerEventToXY(e).x - this.mouse_last_x) +
        parseInt(this.element.css('left'), 10);
      const newY =
        (this.pointerEventToXY(e).y - this.mouse_last_y) +
        parseInt(this.element.css('top'), 10);

      this.element.css({
        left: newX,
        top: newY,
      });

      this.mouse_last_x = this.pointerEventToXY(e).x;
      this.mouse_last_y = this.pointerEventToXY(e).y;
    }
  }

  /**
    * @desc When the user stops dragging the container
  */
  endDrag() {
    if (this.expanded && this.dragging) {
      if (
        this.mouse_start_x === this.mouse_last_x &&
        this.mouse_start_y === this.mouse_last_y
      ) {
        // If an element is clicked & has some content to show
        if (this.clickedElement) {
          if (this.clickedElement.find('.content-detail')[0]) {
            // Show the detailed view.
            // Pass this element as clicked & copy the content to the modal box.
            this.clickedElement.addClass('active');
            this.container
              .addClass('detail-view')
              .removeClass('expanded-view')
              .removeClass('home-view');
            $('.content-detail-container').html(
              this.clickedElement.find('.content-detail').html()
            );

            // Recenter the viewport to have the block
            // positionned on the left side.
            this.element.css({
              left: this.center[0] -
                (this.clickedElement.position().left +
                  this.clickedElement.width() +
                  90),
              top: (this.center[1] / 2) -
                ((this.clickedElement.position().top + this.clickedElement.height()) - 300),
            });
          }
        }
      }
      this.mouse_start_x = this.mouse_last_x = 0;
      this.mouse_start_y = this.mouse_last_y = 0;
      this.dragging = false;
      this.clickedElement = false;
    }
  }

  /**
    * @desc Show the expanded view
  */
  showExpandedView() {
    this.expanded = true;
    this.update();
    this.element.find('.isometric-block').removeClass('active');
    this.container
      .addClass('expanded-view')
      .removeClass('detail-view')
      .removeClass('home-view');
  }

  /**
    * @desc Show the homeview
  */
  showHomeView() {
    this.expanded = false;
    this.update();
    this.container
      .removeClass('expanded-view')
      .removeClass('detail-view')
      .addClass('home-view');

    // Center the grid.
    this.element.css({
      left: 0,
      top: 0,
    });
  }
  /**
    * @desc Reset the center of grid.
  */
  reCenter() {
    this.center[0] = parseInt(this.element.width() / 2, 10);
    this.update();
  }
}


src/house-of-isometric.js
(function () {
  var isometricLayout = function {
    var pointerEventToXY = function (e) {
      var out = { x: 0, y: 0 };
      if (
        e.type == 'touchstart' || e.type == 'touchmove'
        || e.type == 'touchend' || e.type == 'touchcancel'
      ) {
        var touch = e.originalEvent.touches[0] || e.originalEvent.changedTouches[0];
        out.x = touch.pageX;
        out.y = touch.pageY;
      } else if (e.type == 'mousedown' || e.type == 'mouseup' || e.type == 'mousemove' || e.type == 'mouseover' || e.type == 'mouseout' || e.type == 'mouseenter' || e.type == 'mouseleave') {
        out.x = e.pageX;
        out.y = e.pageY;
      }
      return out;
    };

    class Block {
      constructor(grid, element) {
        this.grid = grid;
        this.element = element;
        this.currentcoordinates = JSON.parse(element.attr('data-coordinates'));
      }

      //Return the position of the element from his current coordinates
      getPosition() {
        var x = (this.currentcoordinates[0] - this.currentcoordinates[1]) * this.grid.tile_width / 2 + this.grid.center[0] - this.grid.tile_width / 2;
        var y = (this.currentcoordinates[0] + this.currentcoordinates[1]) * this.grid.tile_height / 2 + this.grid.center[1] - this.grid.tile_height / 2;
        return [x, y];
      }
    }

    class Grid {

      constructor(container, element, tile_width, tile_height) {
        this.blocks = [];
        this.container = container;
        this.element = element;
        this.center = [parseInt(element.width() / 2), 290];
        this.expanded = false;
        this.tile_width = tile_width;
        this.tile_height = tile_height;
        this.dragging = false;
        this.disableClick = false;
        this.mouse_start_x = 0;
        this.mouse_start_y = 0;
        this.mouse_last_x = 0;
        this.mouse_last_y = 0;
        this.clickedElement;
      }

      addBlock(block) {
        this.blocks.push(block);
      }

      disable() {
        this.expanded = !this.expanded;
        if (this.expanded)  {
          this.element.parent('.isometric-layout').addClass('expanded');
        } else {
          this.element.parent('.isometric-layout').removeClass('expanded');
        }
        this.update();
      }

      update() {
        for (const block of this.blocks) {
          var floor_offset = 0;
          if (this.expanded) {
            if (block.element.attr('data-coordinates-expanded') == '[]') {
              block.element.addClass('hidden');
            } else {
              block.currentcoordinates = JSON.parse(block.element.attr('data-coordinates-expanded'));
            }
          } else {
            block.element.removeClass('hidden');
            block.currentcoordinates = JSON.parse(block.element.attr('data-coordinates'));
            floor_offset = parseInt(block.element.attr('data-floor'), 10) * 30;
          }
          block.element.css({ left: block.getPosition()[0], top: block.getPosition()[1] + floor_offset });
        }
      }

      startDrag(e) {
        if (this.expanded && this.dragging === false && this.container.hasClass('expanded-view')) {
          e.preventDefault();
          this.mouse_start_x = this.mouse_last_x = pointerEventToXY(e).x;
          this.mouse_start_y = this.mouse_last_y = pointerEventToXY(e).y;
          this.mouse_start_x = this.mouse_last_x = pointerEventToXY(e).x;
          this.mouse_start_y = this.mouse_last_y = pointerEventToXY(e).y;
          this.dragging = true;
        }
      }

      drag(e) {
        if (this.expanded && this.dragging === true) {
          var new_x = pointerEventToXY(e).x - this.mouse_last_x + parseInt(this.element.css('left'));
          var new_y = pointerEventToXY(e).y - this.mouse_last_y + parseInt(this.element.css('top'));

          // if (new_x > 0) new_x = 0;
          // if (new_y > 0) new_y = 0;
          // if (new_x < -this.element.width()+this.element.parent().width()) new_x = -this.element.width()+this.element.parent().width();
          // if (new_y < -this.element.height()+this.element.parent().height()) new_y = -this.element.height()+this.element.parent().height();

          this.element.css({
            left: new_x,
            top: new_y
          });

          this.mouse_last_x = pointerEventToXY(e).x;
          this.mouse_last_y = pointerEventToXY(e).y;
        }
      }

      endDrag() {
        if (this.expanded && this.dragging) {
          if (this.mouse_start_x === this.mouse_last_x && this.mouse_start_y === this.mouse_last_y) {
            if (this.clickedElement) {
              if (this.clickedElement.find('.content-detail')[0]) {
                this.clickedElement.addClass('active');
                this.container.addClass('detail-view');
                this.container.removeClass('expanded-view');
                this.container.removeClass('home-view');
                $('.content-detail-container').html('');
                $('.content-detail-container').html(this.clickedElement.find('.content-detail').html());
                this.element.css({
                  left: this.center[0] - (this.clickedElement.position().left + this.clickedElement.width() + 90),
                  top: this.center[1] / 2 - (this.clickedElement.position().top + this.clickedElement.height()) + 300
                });
              }
            }
          }
          this.mouse_start_x = this.mouse_last_x = 0;
          this.mouse_start_y = this.mouse_last_y = 0;
          this.dragging = false;
          this.clickedElement = false;
        }
      }

      showExpandedView() {
        this.expanded = true;
        this.update();
        this.element.find('.isometric-block').removeClass('active');
        this.container.addClass('expanded-view');
        this.container.removeClass('detail-view');
        this.container.removeClass('home-view');
      }

      showHomeView() {
        this.expanded = false;
        this.update();
        this.container.removeClass('expanded-view');
        this.container.removeClass('detail-view');
        this.container.addClass('home-view');

        this.element.css({
          left: 0,
          top: 0
        });
      }

      resize() {
        this.center[0] = parseInt(this.element.width() / 2);
        this.update();
      }
    }

    if ($('.isometric-layout')[0])  {
      console.log('here');
      var grid = new Grid($('.isometric-layout'), $('.isometric-grid'), 200, 116);

      $('.isometric-grid .isometric-block').each(function ()  {
        grid.addBlock(new Block(grid, $(this)));
      });

      if (grid.container.hasClass('expanded-view')) {
        grid.expanded = true;
      }
      grid.update();
      grid.container.addClass('ready');

      $('.isometric-grid').on('mousedown touchstart', function (e) {
        grid.startDrag(e);
      });
      $(document).on('mousemove touchmove', function (e) {
        grid.drag(e);
      });
      $(document).on('mouseup touchend', function (e) {
        grid.endDrag();
      });

      $('.isometric-block').on('click touch', function (e) {
        if (!grid.expanded) {
          grid.showExpandedView();
        }
      });

      $('.showExpandedView').on('click touch', function (e) {
        e.preventDefault();
        grid.showExpandedView();
      });

      $('.showHomeView').on('click touch', function (e) {
        e.preventDefault();
        grid.showHomeView();
      });

      $('.isometric-block').on('click touch', function (e) {
        e.preventDefault();
      }).on('mousedown touchstart', function (e) {
        e.preventDefault();
        grid.clickedElement = $(this);
      });

      $('.information-container-detail, .focus-background').on('click touch', function (e) {
        grid.showExpandedView();
      });

      $('.content-detail-container').click(function (event) {
        event.stopPropagation();
      });

      $(window).resize(function () {
        grid.resize();
      });
    }
  }
}());